### ISSUE TITLE
ListView is very slow when debugging.

### ISSUE THREAD

**Issue by vsfeedback** [xamarin/Xamarin.Forms#14218]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/ListView-is-very-slow-when-debugging/1404740)._

---
[severity:It's more difficult to complete my work]
Xamarin.Forms ListView on Android is very slow when debugging, we notice jerky scrolling with 20 items and with more than 50 items it stops responding for 10s of seconds at a time.

If we don't debug then performance is fine.

This makes if very difficult for us to debug issues, because can't attach the debugger or it becomes unusable.

We are following the advice in https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/listview/performance but it doesn't help.

---
#### Feedback Bot on 22/04/2021, 01:50 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Enbi Zhang [MSFT] on 23/04/2021, 01:44 AM: 

<p>Thanks for your feedback!<br>
Could you give a screen video for your issue to help us understanding?</p>
<p><strong>Why do we ask for more info?</strong><br>
We try to reproduce all issues reported with the information provided in the description and comments. When we can’t reproduce the issue, we ask you for more information so we can resolve the issue as quickly and efficiently as possible.<br>
In our <a target='_blank' href="https://aka.ms/vsfeedbackguidelines/#writing-a-good-bug-report-or-feature-suggestion">guidelines</a>, you can get tips on how to provide clear and simple reproducible steps.</p>

#### Feedback Bot on 23/04/2021, 00:14 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Daniel Hughes on 26/04/2021, 05:39 PM: 

<p><a target='_blank' href="https://aka.ms/dc/file?name=Bcc01df543e97405180bb61ebd23ea1f2637550696740642158_ListViewPerformance.zip&amp;tid=cc01df543e97405180bb61ebd23ea1f2637550696740642158">ListViewPerformance.zip</a></p>
<p>I have attached a minimal example that shows the problem:</p>
<p>The list is very simple just three labels in a grid with RecycleElement turned on. On my Google Pixel 1 scrolling is delayed by more than a second and it jumps instead of smoothly scrolling. This with only 50 elements.</p>

---
### Original Solutions
(no solutions)

**Comment by rachelkang** [comment]
I was able to repro this issue with the sample provided 
This also sounds like it's related to https://github.com/xamarin/Xamarin.Forms/issues/11344

I recommend checking out CollectionView if you haven't already!

**Comment by trampster** [comment]
We just switched to CollectionView and can confirm that performance while debugging is greatly improved.

**Comment by ragnar474** [comment]
i must say very laggy is an understatement.
In 4.8.0.1687 we Could get our ListViews to 60fps scroll performance (CachingStrategy="RecycleElementAndDataTemplate").
After 5.0.0.2012 upgrade, all listViews are unusable we are skipping 100+ frames (multiple times) when scrolling.

CollectionView in 5.0, is okay'isch, but still laggy (even with MeasureFirst ).
I gues we wait till MAUI.

BTW.: This occours only while attached with VS.

**Comment by sk1llsh0t** [comment]
I'm experiencing the same thing with the listview. in debug, it lags/logs up the app making it unusable. 

Please fix in a new SR. there is no reason to wait until Maui for this to be resolved.

**Comment by gianisimone** [comment]
I am experiencing the same issue in Xamarin Forms 5.0.0.2012

**Comment by radioactiveman** [comment]
Hi @davidortinau, @jsuarezruiz,

this is yet another issue which should definitely be addressed with a next service release for XF 5.0.
People have to stay on 4.8 just because of this regression.

Could you please prioritize this issue as suggested [here](https://github.com/xamarin/Xamarin.Forms/issues/14205#issuecomment-828450390)? Thanks.

See also: https://github.com/xamarin/Xamarin.Forms/issues/14069

**Comment by AlleSchonWeg** [comment]
Same problem. Garbage collection all the time:

> [Mono] GC_BRIDGE: Complete, was running for 0.06ms
> [Mono] GC_MINOR: (Nursery full) time 12.44ms, stw 13.41ms promoted 1203K major size: 95552K in use: 91900K los size: 21956K in use: 20462K
> [art] Starting a blocking GC Explicit
> [art] Explicit concurrent mark sweep GC freed 7428(328KB) AllocSpace objects, 0(0B) LOS objects, 20% free, 15MB/19MB, paused 402us total 7.093ms
> [Mono] GC_TAR_BRIDGE bridges 6 objects 6 opaque 0 colors 6 colors-bridged 6 colors-visible 6 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.02ms tarjan 0.01ms scc-setup 0.01ms gather-xref 0.01ms xref-setup 0.01ms cleanup 0.02ms
> [Mono] GC_BRIDGE: Complete, was running for 8.29ms
> [Mono] GC_MINOR: (Nursery full) time 13.29ms, stw 14.26ms promoted 1154K major size: 96752K in use: 93075K los size: 25028K in use: 23591K
> [art] Starting a blocking GC Explicit
> [art] Explicit concurrent mark sweep GC freed 2681(119KB) AllocSpace objects, 0(0B) LOS objects, 20% free, 15MB/19MB, paused 408us total 6.888ms
> [Mono] GC_TAR_BRIDGE bridges 2 objects 2 opaque 0 colors 2 colors-bridged 2 colors-visible 2 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.02ms tarjan 0.01ms scc-setup 0.01ms gather-xref 0.01ms xref-setup 0.01ms cleanup 0.02ms
> [Mono] GC_BRIDGE: Complete, was running for 8.10ms
> [Mono] GC_MINOR: (Nursery full) time 12.92ms, stw 13.77ms promoted 1339K major size: 98112K in use: 94440K los size: 29124K in use: 26795K
> [Mono] GC_TAR_BRIDGE bridges 0 objects 0 opaque 0 colors 0 colors-bridged 0 colors-visible 2 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.02ms tarjan 0.01ms scc-setup 0.01ms gather-xref 0.01ms xref-setup 0.01ms cleanup 0.01ms
> [Mono] GC_BRIDGE: Complete, was running for 0.06ms
> [Mono] GC_MINOR: (Concurrent start) time 12.15ms, stw 19.24ms promoted 755K major size: 98896K in use: 95209K los size: 32212K in use: 30070K
> [Mono] GC_MAJOR_CONCURRENT_START: (LOS overflow)
> [art] Starting a blocking GC Explicit
> [art] Explicit concurrent mark sweep GC freed 4443(234KB) AllocSpace objects, 0(0B) LOS objects, 20% free, 15MB/19MB, paused 405us total 8.305ms
> [Mono] GC_TAR_BRIDGE bridges 361 objects 486 opaque 53 colors 361 colors-bridged 360 colors-visible 360 xref 8 cache-hit 0 cache-semihit 0 cache-miss 1 setup 0.05ms tarjan 0.17ms scc-setup 0.04ms gather-xref 0.01ms xref-setup 0.01ms cleanup 0.03ms
> [Mono] GC_BRIDGE: Complete, was running for 14.89ms
> [Mono] GC_MAJOR_CONCURRENT_FINISH: (finishing) time 211.55ms, stw 80.06ms los size: 19908K in use: 5297K
> [Mono] GC_MAJOR_SWEEP: major size: 98144K in use: 92238K
> [art] Starting a blocking GC Explicit
> [art] Explicit concurrent mark sweep GC freed 4317(184KB) AllocSpace objects, 0(0B) LOS objects, 20% free, 15MB/19MB, paused 397us total 6.823ms
> [Mono] GC_TAR_BRIDGE bridges 4 objects 4 opaque 0 colors 4 colors-bridged 4 colors-visible 4 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.03ms tarjan 0.01ms scc-setup 0.01ms gather-xref 0.01ms xref-setup 0.01ms cleanup 0.02ms
> [Mono] GC_BRIDGE: Complete, was running for 8.27ms
> [Mono] GC_MINOR: (Nursery full) time 13.18ms, stw 13.85ms promoted 1239K major size: 98400K in use: 93520K los size: 19908K in use: 8196K
> [art] Starting a blocking GC Explicit
> [art] Explicit concurrent mark sweep GC freed 2632(117KB) AllocSpace objects, 0(0B) LOS objects, 20% free, 15MB/19MB, paused 391us total 6.853ms
> [Mono] GC_TAR_BRIDGE bridges 2 objects 2 opaque 0 colors 2 colors-bridged 2 colors-visible 2 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.03ms tarjan 0.01ms scc-setup 0.01ms gather-xref 0.01ms xref-setup 0.01ms cleanup 0.07ms
> [Mono] GC_BRIDGE: Complete, was running for 8.13ms
> [Mono] GC_MINOR: (Nursery full) time 12.97ms, stw 13.74ms promoted 1321K major size: 99024K in use: 94885K los size: 19908K in use: 11076K
> [Mono] GC_TAR_BRIDGE bridges 0 objects 0 opaque 0 colors 0 colors-bridged 0 colors-visible 2 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.03ms tarjan 0.01ms scc-setup 0.01ms gather-xref 0.01ms xref-setup 0.01ms cleanup 0.01ms
> [Mono] GC_BRIDGE: Complete, was running for 0.06ms
> [Mono] GC_MINOR: (Nursery full) time 12.16ms, stw 13.11ms promoted 1129K major size: 99968K in use: 96053K los size: 19908K in use: 14303K
> [art] Starting a blocking GC Explicit
> [art] Explicit concurrent mark sweep GC freed 6027(273KB) AllocSpace objects, 0(0B) LOS objects, 20% free, 15MB/19MB, paused 398us total 6.850ms
> [Mono] GC_TAR_BRIDGE bridges 4 objects 4 opaque 0 colors 4 colors-bridged 4 colors-visible 4 xref 0 cache-hit 0 cache-semihit 0 cache-miss 0 setup 0.02ms tarjan 0.01ms scc-setup 0.01ms gather-xref 0.02ms xref-setup 0.01ms cleanup 0.02ms
> [Mono] GC_BRIDGE: Complete, was running for 8.21ms
> [Mono] GC_MINOR: (Nursery full) time 12.74ms, stw 13.85ms promoted 1250K major size: 101168K in use: 97343K los size: 20948K in use: 17935K

**Comment by giuseppenovielli** [comment]
Xamarin.Forms 4.8 listview not laggy, also with a complex layout, so @rachelkang check it on Xamarin.Forms v4.8

Thanks!!

**Comment by AlleSchonWeg** [comment]
Hmm,
i just updated VS to 16.10.1 and installed a few windows updates and the problem is gone.

**Comment by gianisimone** [comment]
> Hmm,
> i just updated VS to 16.10.1 and installed a few windows updates and the problem is gone.

Are you using Xamarin Forms 5?

**Comment by AlleSchonWeg** [comment]
> > Hmm,
> > i just updated VS to 16.10.1 and installed a few windows updates and the problem is gone.
> 
> Are you using Xamarin Forms 5?

Yes. I'm not sure what exactly changed the behavior. Windows Update, VS Update or PC restart.

**Comment by dude3133** [comment]
If lags present only in debug mode, then try to disable Hot Reload. It was an issue for me

**Comment by sk1llsh0t** [comment]
> If lags present only in debug mode, then try to disable Hot Reload. It was an issue for me

Thanks for the tip. This fixed it for me. I was able to keep hot reload turned on but changed the hot reload mode from "Changes only" to "Full page". "Full page" hot reload doesn't freeze the debugger/app.

**Comment by rdepin** [comment]
Saw the same issue on Visual Studio for both Mac and Windows.  Disabling hot reload seems to have fixed the issue.  I'm on VS 2019 16.11.1.  BTW, This was also impacting running on a device as well.

**Comment by jfversluis** [comment]
@BretJohnson this seems more like a tooling issue maybe?

**Comment by chrisgull** [comment]
I see a similar symptom with a ListView in the latest (XF 2196) on Android. 
With 2196 I can observe the Live Visual Tree, and I see that the ListView runs amok, adding ViewCells in perpetuity (stopped watching at 100,000). 
My list has five (5) items.

RetainElement eliminates the issue but is slowish in performance.
Without an ItemTemplate the issue does not appear.
Rolling back to XF 1931 fixes the problem.

I hope this helps.

**Comment by Sebastian1989101** [comment]
@jfversluis @davidortinau any chance this will get fixed? This is a major issue because many apps uses lists. There are even PR's that focus on this problem. The quality of Xamarin is utterly bad lately... I would even say it's unuseable at this point with all this bugs open and even critical ones like this one keept open for months...

**Comment by jfversluis** [comment]
Hey @Sebastian1989101 ! I see you copy/pasted this comment in multiple places ([here](https://github.com/xamarin/Xamarin.Forms/issues/14069#issuecomment-974730923) and [here](https://github.com/xamarin/Xamarin.Forms/issues/11344#issuecomment-974730841)). If you don't mind I will address this only once :)

I understand the situation you're in. I hope that you have been seeing that we are putting a lot of effort in pushing out service releases (also see #14205 for some context). Together with that we have been bringing down the number of PRs and issues significantly. I wish I could snap my fingers and go through all of them today, unfortunately that is not possible.

There are a lot of issues that are important to a lot of different people, including this one. Also here, I wish I could unblock everyone in an instant, but I can't. I am doing the best I can going through them and fixing what we can while trying to keep an eye on regressions.

I can go on and explain how we got here, but I'd rather focus on how we go from here. We want to see this fixed and happy customers just as much as you.

For now we are releasing a service release every month. You can see the issues and PRs that have our focus on the [project boards](https://github.com/xamarin/Xamarin.Forms/projects) that I create for each month. I checked and noticed this one was one it, the other ones you mentioned were not so I fixed that.

It's hard to promise anything in terms of when we fix what, because there is a lot of things going on and everyone feels their issue is the most important one, which makes sense. 

If something is really important to you and blocking you, the only thing I can ask is if we can work together on this, because I can use some help. If you're willing and able, it would probably speed up the process if you could pinpoint more precisely where in the code something seems to go wrong or with which PR it was introduced. If you could provide a PR that fixes it that would of course be totally amazing and very much appreciated.

It's maybe not the answer you'd like at this moment, but the only thing I can do right now is be open, honest and transparent about what is going on and kindly ask for any assistance you can offer. We're doing the best we can to make your Forms experience the best we can while it lasts :)

I'd love not to litter any issues with discussions like these, so if you still have any concerns please follow up through any channel of your choosing either private or public. I would suggest email for private and GitHub discussions for public.

Thanks!

**Comment by chrisgull** [comment]
> Hey @Sebastian1989101 ! I see you copy/pasted this comment in multiple places ([here](https://github.com/xamarin/Xamarin.Forms/issues/14069#issuecomment-974730923) and [here](https://github.com/xamarin/Xamarin.Forms/issues/11344#issuecomment-974730841)). If you don't mind I will address this only once :)
> 
> I understand the situation you're in. I hope that you have been seeing that we are putting a lot of effort in pushing out service releases (also see #14205 for some context). Together with that we have been bringing down the number of PRs and issues significantly. I wish I could snap my fingers and go through all of them today, unfortunately that is not possible.
> 
> There are a lot of issues that are important to a lot of different people, including this one. Also here, I wish I could unblock everyone in an instant, but I can't. I am doing the best I can going through them and fixing what we can while trying to keep an eye on regressions.
> 
> I can go on and explain how we got here, but I'd rather focus on how we go from here. We want to see this fixed and happy customers just as much as you.
> 
> For now we are releasing a service release every month. You can see the issues and PRs that have our focus on the [project boards](https://github.com/xamarin/Xamarin.Forms/projects) that I create for each month. I checked and noticed this one was one it, the other ones you mentioned were not so I fixed that.
> 
> It's hard to promise anything in terms of when we fix what, because there is a lot of things going on and everyone feels their issue is the most important one, which makes sense.
> 
> If something is really important to you and blocking you, the only thing I can ask is if we can work together on this, because I can use some help. If you're willing and able, it would probably speed up the process if you could pinpoint more precisely where in the code something seems to go wrong or with which PR it was introduced. If you could provide a PR that fixes it that would of course be totally amazing and very much appreciated.
> 
> It's maybe not the answer you'd like at this moment, but the only thing I can do right now is be open, honest and transparent about what is going on and kindly ask for any assistance you can offer. We're doing the best we can to make your Forms experience the best we can while it lasts :)
> 
> I'd love not to litter any issues with discussions like these, so if you still have any concerns please follow up through any channel of your choosing either private or public. I would suggest email for private and GitHub discussions for public.
> 
> Thanks!

Gerald, If there's anything I can contribute on this (repro solution perhaps) to help you track down the problem then please let me know. I have settled for XF 1931 for now, but would of course love to be on the latest. My app is released across all three platforms and seems to work okay, but I'm honestly not sure what I am missing out on.
Best,
Lars

**Comment by jfversluis** [comment]
Thanks Lars! For the issues that are important to you see what you can do. Reproductions always definitely help :) If from there you have some idea where in the Forms code it might be going on and you can poke around in that, even better! Of you can then fix it, that is the best! :D

But any of these stages is much appreciated!

**Comment by radoczsandor** [comment]
I still have this issue in debug mode.
